web/src/message_fetch.ts

```diff
@@ -475,39 +577,80 @@ export function maybe_load_older_messages(opts) {
         }
 
         opts.cont = () =>
-            setTimeout(() => maybe_load_older_messages(opts), consts.catch_up_backfill_delay);
+            setTimeout(() => {
+                maybe_load_older_messages(opts);
+            }, consts.catch_up_backfill_delay);
+    }
+    if (opts.msg_list) {
+        do_backfill({
+            msg_list: opts.msg_list,
+            cont() {
+                if (opts.cont) {
+                    opts.cont();
+                }
+            },
+            num_before: opts.recent_view
+                ? consts.recent_view_fetch_more_batch_size
+                : consts.narrowed_view_backward_batch_size,
+        });
+    } else {
+        do_backfill({
+            msg_list_data: opts.msg_list_data,
+            cont() {
+                if (opts.cont) {
+                    opts.cont();
+                }
+            },
+            num_before: opts.recent_view
+                ? consts.recent_view_fetch_more_batch_size
+                : consts.narrowed_view_backward_batch_size,
+        });
     }
-
-    do_backfill({
-        ...opts,
-        num_before: opts.recent_view
-            ? consts.recent_view_fetch_more_batch_size
-            : consts.narrowed_view_backward_batch_size,
-    });
 }
 
-export function do_backfill(opts) {
-    const msg_list_data = opts.msg_list_data ?? opts.msg_list.data;
+export function do_backfill(
+    opts: {
+        num_before: number;
+        cont?: () => void;
+    } & (
+        | {msg_list: message_lists.MessageList; msg_list_data?: undefined}
+        | {msg_list?: undefined; msg_list_data: MessageListData}
```

Comment:
Since `load_messages` alreads checks which one of the two is defined, we should be able to combine them?

Reply:
I don't like how this introduces so much new code, if there is no way around it, we should be able to store parameters in a variable and use it twice.

Reply:
I separated this as per the [suggestion](https://github.com/zulip/zulip/pull/30509#discussion_r1649454634) because we needed to assert the msg_list_data to convince TypeScript in the earlier iteration of the PR. Combining this will re-introduce the assert logic.

Also combining it here only will be like 
```
export function do_backfill(
    opts: {
        num_before: number;
        cont?: () => void;
        msg_list?: message_lists.MessageList | undefined
        msg_list_data?: MessageListData | undefined
    })
    
  ```
  but again there is no good way to convince TypeScript that either of them must be defined.
  
  One thing we can do is passing `msg_list_data` explicitly (as mentioned in that comment), this will remove the extra conditionals in this function.